File: src/services/anime/watchlist.ts
import { redisCache } from '@/services/redis/cache'

// Redis key prefix for watchlists
const WATCHLIST_KEY_PREFIX = 'watchlist:user:'

export async function getUserWatchlist(userId: string): Promise<number[]> {
  try {
    const redisKey = `${WATCHLIST_KEY_PREFIX}${userId}`
    const watchlistItems = await redisCache.getSetMembers(redisKey)
    return watchlistItems.map((item) => parseInt(item, 10)).filter((id) => !isNaN(id))
  } catch (error) {
    console.error(`Error getting watchlist for user ${userId}:`, error)
    return []
  }
}

export async function addToWatchlist(
  userId: string,
  animeId: number
): Promise<{ success: boolean; message: string }> {
  try {
    const redisKey = `${WATCHLIST_KEY_PREFIX}${userId}`

    // Check if anime is already in watchlist
    const isAlreadyInWatchlist = await redisCache.isInSet(redisKey, animeId)
    if (isAlreadyInWatchlist) {
      return { success: false, message: 'Anime already in your watchlist.' }
    }

    // Add anime to watchlist
    await redisCache.addToSet(redisKey, animeId)

    // Set a reasonable TTL for the watchlist (30 days)
    await redisCache.expire(redisKey, 30 * 24 * 60 * 60)

    return { success: true, message: 'Anime added to your watchlist.' }
  } catch (error) {
    console.error(`Error adding anime ${animeId} to watchlist for user ${userId}:`, error)
    return { success: false, message: 'Failed to add anime to watchlist.' }
  }
}

export async function removeFromWatchlist(
  userId: string,
  animeId: number
): Promise<{ success: boolean; message: string }> {
  try {
    const redisKey = `${WATCHLIST_KEY_PREFIX}${userId}`

    // Check if anime is in watchlist
    const isInWatchlist = await redisCache.isInSet(redisKey, animeId)
    if (!isInWatchlist) {
      return { success: false, message: 'Anime not found in your watchlist.' }
    }

    // Remove anime from watchlist
    await redisCache.removeFromSet(redisKey, animeId)

    return { success: true, message: 'Anime removed from your watchlist.' }
  } catch (error) {
    console.error(`Error removing anime ${animeId} from watchlist for user ${userId}:`, error)
    return { success: false, message: 'Failed to remove anime from watchlist.' }
  }
}

File: go/internal/services/anilist/watchlist.go
package anilist

import (
	"context"
	"strconv"
	"time"

	"discord-anime-bot/internal/services/redis"
)

const watchlistKeyPrefix = "watchlist:user:"


// AddToWatchlist adds an anime to a user's watchlist
func AddToWatchlist(userID string, animeID int) (string, error) {
	ctx := context.Background()
	redisKey := watchlistKeyPrefix + userID

	// Check if anime is already in watchlist
	isInWatchlist, err := redis.SetIsMember(ctx, redisKey, animeID)
	if err != nil {
		return "", err
	}
	if isInWatchlist {
		return "Anime already in your watchlist.", nil
	}

	// Add anime to watchlist
	if err := redis.SetAdd(ctx, redisKey, animeID); err != nil {
		return "", err
	}

	// Set TTL (30 days)
	if err := redis.Expire(ctx, redisKey, 30*24*time.Hour); err != nil {
		return "", err
	}

	return "Anime added to your watchlist.", nil
}

// RemoveFromWatchlist removes an anime from a user's watchlist
func RemoveFromWatchlist(userID string, animeID int) (string, error) {
	ctx := context.Background()
	redisKey := watchlistKeyPrefix + userID

	// Check if anime is in watchlist
	isInWatchlist, err := redis.SetIsMember(ctx, redisKey, animeID)
	if err != nil {
		return "", err
	}
	if !isInWatchlist {
		return "Anime not found in your watchlist.", nil
	}

	// Remove anime from watchlist
	if err := redis.SetRemove(ctx, redisKey, animeID); err != nil {
		return "", err
	}

	return "Anime removed from your watchlist.", nil
}

// GetUserWatchlist returns a user's watchlist
func GetUserWatchlist(userID string) ([]int, error) {
	ctx := context.Background()
	redisKey := watchlistKeyPrefix + userID

	// Get all members of the set
	members, err := redis.SetMembers(ctx, redisKey)
	if err != nil {
		return []int{}, nil // Return empty slice on error
	}

	// Convert string members to integers
	var animeIDs []int
	for _, member := range members {
		if animeID, err := strconv.Atoi(member); err == nil {
			animeIDs = append(animeIDs, animeID)
		}
	}

	return animeIDs, nil
}
